book/embeds.md

```diff
@@ -88,147 +73,193 @@ def request(url, top_level_url, payload=None):
-`content-type` and assume that if it starts with `text` or contains
-`javascript`, the content is
-`utf8`, and otherwise return it as un-decoded data:
+Note that I _didn't_ add a `decode` call when we read the body; that's
+because the body might actually be binary data, and we want to return
+that binary data directly to the browser. Now, every existing call to
+`request`, which wants textual data, needs to `decode` the response.
+For example, in `load`, you'll want to do something like this:
 
-``` {.python}
-def request(url, top_level_url, payload=None):
-    # ...
-    content_type = headers.get(
-        'content-type',
-        'application/octet-stream')
-    if content_type.startswith("text") or \
-        content_type.find('javascript') >= 0:
-        body = response.read().decode("utf8")
-    else:
-        body = response.read()
-    # ...
-    return headers, body
+``` {.python replace=Tab/Frame}
+class Tab:
+    def load(self, url, body=None):
+        # ...
+        headers, body = request(url, self.url, payload=body)
+        body = body.decode("utf8")
+        # ...
 ```
 
-Now define a method that decodes a response body that we know is an image
-(even if we don't know its format).^[Interestingly, to make it work for our toy
-browser we don't need to consult `content-type`. That's because Skia already
-auto-detects the image format by peeking at the first few bytes of the binary
-data, which varies for each image format.][^maybe-decode]
+Make sure to make this change everywhere in your browser that you call
+`request`, including inside `XMLHttpRequest_send` and in several other
+places in `request`. When we download images, however, we _won't_ call
+`decode`, and just use the binary data directly:
+
+``` {.python replace=tab/frame}
+def download_image(image_src, tab):
+    image_url = resolve_url(image_src, tab.url)
+    assert tab.allowed_request(image_url), \
+        "Blocked load of " + image_url + " due to CSP"
+    header, body = request(image_url, tab.url)
+    data = skia.Data.MakeWithoutCopy(body)
+    img = skia.Image.MakeFromEncoded(data)
+    assert img, "Failed to recognize image format for " + image_url
+    return body, img
+```
+
+Let's look at the steps between `request` and `return` carefully.
+First, the requested data is turned into a Skia `Data` object using
+the `MakeWithoutCopy` method. Then that `Data` is used to create an
+`Image` using `MakeFromEncoded`.
+
+`MakeWithoutCopy` means that the `Data` object just stores a reference
+to the existing `body` and doesn't own that data. That's essential,
+because encoded image data can be large---maybe megabytes---and
+copying that data wastes memory and time. But that means that the
+`data` is invalid if `body` is ever garbage-collected, so we return
+the `body` from `download_image` and need to make sure to store it
+somewhere for at least as long as we're using the image.[^memoryview]
+
+[^memoryview]: This is a bit of a hack. Perhaps a better solution
```

Comment:
I don't quite understand why it's a hack, can you explain?

Reply:
Uhh, well, IDK, I think a "nice" solution wouldn't involve working around garbage collector weaknesses. In C++ I would "move" the request object into the Skia Data, and or writing directly to a Data would be nicer. But perhaps what you really mean is—let's explain the footnote a bit better so people aren't confused. Will do.

Reply:
ok sounds good